### DISCUSSION TITLE
Validation accuracy oscillation

### DISCUSSION THREAD

**Discussion by Mousa-saleh** [neuropoly/axondeepseg#572] [Show and tell]
Hello all,

I have 13 images(1024*1370), I trained the model with 10 images (60 patches ). 42 patches for training and 18 for validation. I got high fluctuation in the validation accuracy. 
I did multiple runs (attached)., each run with different augmentation and dropout. I kept other config parameters as shown below 

```
"n_classes": 3,  
  "thresholds": [0, 0.2, 0.8],    
  "trainingset_patchsize": 512,   
  "trainingset": "TEM",  # Name of the training set.
  "batch_size": 2,  
  "epochs":1000,
  "checkpoint_period": 5, 
  "checkpoint": None, 

# Network architecture parameters:     
  "depth": 4,  
  "convolution_per_layer": [2, 2, 2, 2],  
  "size_of_convolutions_per_layer": [[5, 5], [3, 3], [3, 3], [3, 3]], 
  "features_per_convolution": [[[1, 16], [16, 16]], [[16, 32], [32, 32]], [[32, 64], [64, 64]], [[64, 128], [128, 128]]],  
  "downsampling": "convolution",


# Learning rate parameters:  
"learning_rate": 1e-4, .  
  "learning_rate_decay_activate": True,  
  "learning_rate_decay_period": 24000,  
  "learning_rate_decay_type": "polynomial", 
  "learning_rate_decay_rate": 0.99,  

# Batch normalization parameters: 
batch_norm_activate": True,  
  "batch_norm_decay_decay_activate": True,  
  "batch_norm_decay_starting_decay": 0.7,  
  "batch_norm_decay_ending_decay": 0.9,  
  "batch_norm_decay_decay_period": 16000,  


# Weighted cost parameters:    
  "weighted_cost-activate": True,  
  "weighted_cost-balanced_activate": True,
  "weighted_cost-balanced_weights": [1, 1.15, 3],  
  "weighted_cost-boundaries_sigma": 2,  
  "weighted_cost-boundaries_activate": False,  

``
[ADS_results - Copy.docx](https://github.com/neuropoly/axondeepseg/files/7329353/ADS_results.-.Copy.docx)
`
Thanks,

**Comment by Stoyan-I-A** [comment]
The link to your word document isn't working because of the GitHub formating. Here's the file for whoever reads this later:
[ADS_results.-.Copy.docx](https://github.com/neuropoly/axondeepseg/files/7333629/ADS_results.-.Copy.docx)
 
@mariehbourget Have you seen something like this during the training?

@Mousa-saleh The validation accuracies are indeed odd. How well does the trained model perform on your images? (Do you have a sample segmented image?)

**Reply by Mousa-saleh** [reply]
I think the main reason for this oscillation is that I am using batch size of 2. I increased it to 8 and getting good results as in the below picture. 

![image](https://user-images.githubusercontent.com/85881282/137063837-7d5b6e5d-c724-4c52-91a0-4463af4f9216.png)

**Reply by Stoyan-I-A** [reply]
Good catch! It seems like the correct explanation.

**Comment by Mousa-saleh** [comment]
Hello guys, 
I got a good model during training(attached) but when I used the testing set as the validation set, I got oscillations again.(attached).
The total number of images I had is 13, 7 for training, 3 for validation and 3 for testing.
Do you think I should have used more images during training or the total number of images itself is  not enough especially that the images have different contrast due to staining.

Thanks,

![image](https://user-images.githubusercontent.com/85881282/137584886-4ddc4492-14ac-43a1-9a1a-4675bebff205.png)
![image](https://user-images.githubusercontent.com/85881282/137584904-f72d8c21-ae35-474a-8280-3a2469a9d1d4.png)

**Reply by Stoyan-I-A** [reply]
> the images have different contrast due to staining

This can affect the results quite a bit. I wonder if some kind of pre-processing could be used to normalise the contrast of every image. This could be especially useful if the other microscopy images you want to segment also have a contrast that slightly differs from your training set.

I suggest to include as many "variations of contrast" as possible in your training set,.
As for the dataset size, we usually only use one image for testing. So you could try it with 8 images for training, 4 for validation and 1 for testing.
